Add `use` declaration re-ordering

studoot opened the pull request:
This partially addresses #298, adding sorting of contiguous runs of `use` declarations depending on the value and type of path referenced. As an example (taken from the added integration test files), the following code

```rust
use std::str;
use std::cmp::{d, c, b, a};
use std::ddd::aaa;
use std::ddd::{d as p, c as g, b, a};
// This comment should stay with `use std::ddd:bbb;`
use std::ddd::bbb;
```

gets transformed to 

```rust
use std::cmp::{a, b, c, d};
use std::ddd::{a, b, c as g, d as p};
use std::ddd::aaa;
// This comment should stay with `use std::ddd:bbb;`
use std::ddd::bbb;
use std::str;
```

The `reorder_imports` config item has changed from a boolean to {None, Lines, Items, LinesAndItems}, to allow combinations of use declaration lines and use declaration items to be selected.

I would have to say that this isn't a finished item - there are several areas that I'm not 100% happy with:

- Any 'missing spans' before the run of `use` declarations remain at the start of the run, even after the declarations are reordered. I can see situations where a comment is expected to stay attached to the first `use` in a run. I'm thinking situations like:

    ```rust
// This is a file header comment

// We need to use the filesystem
use std::fs;
use std::cmp;
    ```

    which will currently become

    ```rust
// This is a file header comment

// We need to use the filesystem
use std::cmp;
use std::fs;
    ```

- Similarly, any end of line comment after a `use` will stay with the code after that use:

    ```rust
use std::fs; // We need to use the filesystem
use std::cmp;
    ```

    which will currently become

    ```rust
// We need to use the filesystem
use std::cmp;
use std::fs;
    ```

- Contents of import lists are currently ignored when comparing imports, meaning that `use a::b::{c, d};` is considered equal to `use a::b::{e, f};` (but that is relatively easy to implement).

- Comparison functions for AST items such as paths etc are clustered at the top of `visitor.rs`. Is there either a) a preferable place to put them, or b) a preferable way to implement them?!

- Looking at the code again, some refactoring would be beneficial so that the `use` declarations in functions were processed the same as those in modules (that would mean changing the code around the call to `visit_item` in `visit_stmt`)

Anyway - it's a start...

nrc commented:
Also todo for later is that I would expect imports to be squished together, e.g.,

```
use std::ddd::aaa;
use std::ddd::{d as p, c as g, b, a};
// This comment should stay with `use std::ddd:bbb;`
use std::ddd::bbb;
```

becomes

```
use std::ddd::{aaa, a, b, c as g, d as p};
// This comment should stay with `use std::ddd:bbb;`
use std::ddd::bbb;
```

Note the second import stays separate due to the comment.

nrc commented:
> Contents of import lists are currently ignored when comparing imports, meaning that use a::b::{c, d}; is considered equal to use a::b::{e, f}; (but that is relatively easy to implement).

Could you file an issue for this and add a `FIXME(#nn)` to the code please?

nrc commented:
> Comparison functions for AST items such as paths etc are clustered at the top of visitor.rs. Is there either a) a preferable place to put them, or b) a preferable way to implement them?!

I would put them in imports.rs, I guess the more generic stuff (e.g., Paths) should be somewhere else, but we can deal with that if they become more widely used.

nrc commented on src/config.rs:
```diff
@@ -170,6 +170,27 @@ configuration_option_enum! { WriteMode:
+    Items,
+    // Order `use` statements by their path and items within `use` statements by their name.
+    LinesAndItems
+}
```
I would prefer to add another option, rather than have a single enum like this - this way lies combinatorial explosion!

studoot replied:
Absolutely - I had a look to see if there was a way that a list of enumerations could be supplied rather than just one (something like `reorder_imports = [ "Lines", "Imports" ]` or `reorder_imports = "Lines,Imports"`), but that was looking a bit involved, so I'll change to separate options.

nrc commented on src/visitor.rs:
```diff
@@ -31,6 +32,63 @@ macro_rules! source {
+        _ => false,
+    }
+}
+fn compare_use_items(a: &ast::Item, b: &ast::Item) -> Option<Ordering> {
```
nit: newline between functions

nrc commented on src/visitor.rs:
```diff
@@ -502,8 +560,21 @@ impl<'a> FmtVisitor<'a> {
-        for item in &m.items {
-            self.visit_item(&item);
+        let mut items_left: &[ptr::P<ast::Item>] = &m.items;
+        while !items_left.is_empty() {
```
Could you add a comment explaining what you are doing here please?

nrc commented on src/visitor.rs:
```diff
@@ -547,6 +618,30 @@ impl<'a> FmtVisitor<'a> {
         self.format_missing(filemap.end_pos);
     }
 
+    fn format_imports(&mut self, use_items: &[ptr::P<ast::Item>]) {
```
Can you factor this code out and into imports.rs

nrc commented:
This looks great, thanks for the PR! A few very minor comments to address.

studoot commented:
- [X] Sorting on import item lists *was* simple enough, so I've added that to this PR...
- [X] Moved `format_imports` and `format_import` to `imports.rs`
- [X] Added comment to `walk_mod_items`
- [X] Changed reorder_imports options to two separate boolean options

nrc commented:
Thank you!
